File: mobilewebview/android/src/org/mobilewebview/OriginUtils.java
package org.mobilewebview;

import java.net.URI;
import java.util.List;

final class OriginUtils {
    private OriginUtils() {
    }

    static String extractOrigin(String url) {
        if (url == null || url.isEmpty()) {
            return "";
        }

        try {
            URI uri = URI.create(url);
            String scheme = uri.getScheme();
            String host = uri.getHost();
            int port = uri.getPort();

            if (scheme == null || host == null) {
                return "";
            }

            StringBuilder origin = new StringBuilder();
            origin.append(scheme).append("://").append(host);

            // Only include port if non-standard.
            if (port != -1 && port != 80 && port != 443) {
                origin.append(":").append(port);
            }

            return origin.toString();
        } catch (RuntimeException e) {
            return "";
        }
    }

    static boolean isOriginAllowed(String origin, List<String> allowedOrigins) {
        if (allowedOrigins == null || allowedOrigins.isEmpty()) {
            return false;
        }

        if (origin == null || origin.isEmpty()) {
            return false;
        }

        if (allowedOrigins.contains(origin) || allowedOrigins.contains("*")) {
            return true;
        }

        for (String pattern : allowedOrigins) {
            if (pattern != null && pattern.contains("*")) {
                String regex = pattern.replace(".", "\\.").replace("*", ".*");
                if (origin.matches(regex)) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: mobilewebview/src/common/origin_utils.cpp
#include "origin_utils.h"
#include <QUrl>
#include <QStringList>
#include <QRegularExpression>

QString extractOrigin(const QUrl &url)
{
    if (!url.isValid() || url.scheme().isEmpty() || url.host().isEmpty()) {
        return QString();
    }
    
    QString origin = url.scheme() + QStringLiteral("://") + url.host();
    int port = url.port();
    if (port != -1 && port != 80 && port != 443) {
        origin += QLatin1Char(':') + QString::number(port);
    }
    return origin;
}

bool isOriginAllowed(const QString &origin, const QStringList &allowedOrigins)
{
    // If no allowlist is set, reject all
    if (allowedOrigins.isEmpty()) {
        return false;
    }
    
    // Empty origin is never allowed
    if (origin.isEmpty()) {
        return false;
    }
    
    // Check for exact match first (including "*" wildcard)
    if (allowedOrigins.contains(origin) || allowedOrigins.contains(QLatin1String("*"))) {
        return true;
    }
    
    // Check for wildcard patterns (e.g., "*.example.com" or "*://example.com")
    for (const QString &pattern : allowedOrigins) {
        if (pattern.contains(QLatin1Char('*'))) {
            // Convert wildcard pattern to regex
            QString regexPattern = QRegularExpression::escape(pattern);
            regexPattern.replace(QStringLiteral("\\*"), QStringLiteral(".*"));
            regexPattern = QLatin1String("^") + regexPattern + QLatin1String("$");
            
            QRegularExpression regex(regexPattern, QRegularExpression::CaseInsensitiveOption);
            if (regex.match(origin).hasMatch()) {
                return true;
            }
        }
    }
    
    return false;
}
